internal/controller/azurevalidator_controller.go

```diff
@@ -68,8 +63,8 @@ func (r *AzureValidatorReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
-	// Ensure all four Azure env vars are set.
-	if err := r.configureAzureAuth(validator.Spec.Auth, req.Namespace, l); err != nil {
+	// Override auth data in spec with auth data from Secret if applicable.
+	if validator.Spec.Auth, err = r.authFromSecret(validator.Spec.Auth, req.Namespace, l); err != nil {
```

Comment:
Can you clarify? I was thinking that doing validation this way does work because the validation happens in `validateAuth` which is on both code paths, the one from the controller in continuous mode (`Reconcile`) and the one from `validatorctl` in direct mode (`Validate`).

![image](https://github.com/user-attachments/assets/b6908147-627c-471a-83a5-cddf208d15cf)

There is the flaw in this design where the validation code doesn't know whether the data came from the secret or from the spec, but I wasn't worrying about that for now.

Reply:
Yeah, I was primarily concerned about the fact that the values from the secret are only set if they're present and if they're not, the code just silently continues.

Ideally we would return an error that specifically indicates an invalid secret. As in, when a secret is indicated in the spec and a required env var is missing from the secret, the failure message should indicate that clearly.

Reply:
Ah that's fair. I'l rework it a bit so we do validation on the secret and on the inline auth. A bit of repeated code is worth it for good user experience.

Reply:
I think you can avoid repeating code by just returning errors from `authFromSecret` for any required var where `ok==false`.